File: Angular_Frontend/WorkBridgeAngular/src/app/auth/services/storage.service.ts
import { Injectable } from '@angular/core';
import { LoginResponseModel } from '../models/login-response.model';
import { CryptoUtil } from '../../utils/crypto.util';


export const KEYS = {
  TOKEN: 'wb_token',
  LOGIN_USER: 'wb_login_user',

  USER_PROFILE: 'wb_user_profile',
  COMPANY_PROFILE: 'wb_company_profile'
};

@Injectable({
  providedIn: 'root',
})
export class StorageService {


  constructor() { }

  // ==========================
  // Login Session
  // ==========================

  saveSession(data: LoginResponseModel): void {

    localStorage.setItem(
      KEYS.TOKEN,
      CryptoUtil.encrypt(data.token)
    );

    localStorage.setItem(
      KEYS.LOGIN_USER,
      CryptoUtil.encrypt(JSON.stringify(data))
    );

  }

  getToken(): string | null {

    const raw = localStorage.getItem(KEYS.TOKEN);

    return raw ? CryptoUtil.decrypt(raw) : null;

  }

  getUser(): LoginResponseModel | null {

    const raw = localStorage.getItem(KEYS.LOGIN_USER);

    if (!raw) return null;

    try {

      const json = CryptoUtil.decrypt(raw);

      return json ? JSON.parse(json) : null;

    } catch {

      return null;

    }

  }

  getRole(): string | null {

    return this.getUser()?.role ?? null;

  }

  getUserId(): number | null {

    return this.getUser()?.userId ?? null;

  }

  getProfileId(): number | null {

    return this.getUser()?.profileId ?? null;

  }

  isLoggedIn(): boolean {

    return !!this.getToken();

  }

  clearSession(): void {

    Object.values(KEYS).forEach(key =>
      localStorage.removeItem(key)
    );

  }

  // ==========================
  // Generic Methods
  // ==========================

  saveData(key: string, data: any): void {

    localStorage.setItem(
      key,
      CryptoUtil.encrypt(JSON.stringify(data))
    );

  }

  getData<T>(key: string): T | null {

    const raw = localStorage.getItem(key);

    if (!raw) return null;

    try {

      const json = CryptoUtil.decrypt(raw);

      return json ? JSON.parse(json) : null;

    } catch {

      return null;

    }

  }

  removeData(key: string): void {

    localStorage.removeItem(key);

  }

  // ==========================
  // User Profile
  // ==========================

  saveUserProfile(profile: any): void {

    this.saveData(KEYS.USER_PROFILE, profile);

  }

  getUserProfile<T>(): T | null {

    return this.getData<T>(KEYS.USER_PROFILE);

  }

  // ==========================
  // Company Profile
  // ==========================

  saveCompanyProfile(profile: any): void {

    this.saveData(KEYS.COMPANY_PROFILE, profile);

  }

  getCompanyProfile<T>(): T | null {

    return this.getData<T>(KEYS.COMPANY_PROFILE);

  }


  // ==========================
  // JWT Expiration
  // ==========================

  isTokenExpired(): boolean {

    const token = this.getToken();

    if (!token) {
      return true;
    }

    try {

      const payload = JSON.parse(
        atob(token.split('.')[1])
      );

      const exp = payload.exp;

      if (!exp) {
        return false;
      }

      return Date.now() >= exp * 1000;

    } catch {

      return true;

    }

  }



}

File: Flutter_Frontend/work_bridge_flutter/lib/services/storage_service.dart
import 'dart:convert';

import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:work_bridge_flutter/auth/request/user_request.dart';
import 'package:work_bridge_flutter/auth/response/login_response.dart';

class StorageKeys {
  StorageKeys._();

  static const token = 'wb_token';
  static const user = 'wb_user';
}

class StorageService {
  StorageService(this._storage);

  final FlutterSecureStorage _storage;

  // ── Write ────────────────────────────────────────────
  Future<void> saveSession(LoginResponse data) async {
    await _storage.write(key: StorageKeys.token, value: data.token);
    await _storage.write(
      key: StorageKeys.user,
      value: jsonEncode(data.toJson()),
    );
  }

  // ── Read ─────────────────────────────────────────────
  Future<String?> getToken() => _storage.read(key: StorageKeys.token);

  Future<LoginResponse?> getUser() async {
    final raw = await _storage.read(key: StorageKeys.user);
    if (raw == null) return null;
    try {
      return LoginResponse.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } catch (_) {
      return null;
    }
  }

  Future<UserRole?> getRole() async => (await getUser())?.role;

  Future<bool> isLoggedIn() async => (await getToken()) != null;

  // ── Clear ────────────────────────────────────────────
  Future<void> clearSession() async {
    await _storage.delete(key: StorageKeys.token);
    await _storage.delete(key: StorageKeys.user);
  }

  // ── Generic (mirrors saveData/getData/removeData) ────
  Future<void> saveData(String key, Map<String, dynamic> data) =>
      _storage.write(key: key, value: jsonEncode(data));

  Future<Map<String, dynamic>?> getData(String key) async {
    final raw = await _storage.read(key: key);
    if (raw == null) return null;
    try {
      return jsonDecode(raw) as Map<String, dynamic>;
    } catch (_) {
      return null;
    }
  }

  Future<void> removeData(String key) => _storage.delete(key: key);
}
